### PR TITLE
 On partial uninit error point at where we need init

### LINKED ISSUE #97956: Possibly-uninitialized error message doesn't indicate where value isn't initialized
Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d0d96c280bcde85e4d524178fefca741

```rust
let pkg;
if true {
    pkg = 1;
} else {
    // I forgot to initialize pkg!
}
drop(pkg);
```

The current output is:

```
error[[E0381]](https://doc.rust-lang.org/stable/error-index.html#E0381): use of possibly-uninitialized variable: `pkg`
 --> src/main.rs:8:10
  |
8 |     drop(pkg);
  |          ^^^ use of possibly-uninitialized `pkg`
```

Ideally the output should look like:

```
error[[E0381]](https://doc.rust-lang.org/stable/error-index.html#E0381): use of possibly-uninitialized variable: `pkg`
 --> src/main.rs:8:10
  |
8 |     drop(pkg);
  |          ^^^ use of possibly-uninitialized `pkg`
  |
  | hint: the value is not assigned in the branch starting on src/main.rs:5:
  |
5 |     } else {
  |
```

Because it can be difficult to figure out exactly which branch is failing to initialize a value, especially if it's not as simple as one `if` as above. For example, there may be a whole tree of conditionals between the `let` and the use, where only one sub-branch fails to initialize the variable, but spotting that amidst all the code isn't easy.

### PR DESCRIPTION
**Pull request by estebank**
When a binding is declared without a value, borrowck verifies that all
codepaths have *one* assignment to them to initialize them fully. If
there are any cases where a condition can be met that leaves the binding
uninitialized or we attempt to initialize a field of an uninitialized
binding, we emit E0381.

We now look at all the statements that initialize the binding, and use
them to explore branching code paths that *don't* and point at them. If
we find *no* potential places where an assignment to the binding might
be missing, we display the spans of all the existing initializers to
provide some context.

Fix https://github.com/rust-lang/rust/issues/97956.

### CODE REVIEW

File: src/test/ui/nll/match-cfg-fake-edges.stderr [hunk 1]
```diff
@@ -1,8 +1,11 @@
-error[E0381]: use of possibly-uninitialized variable: `x`
+error[E0381]: used binding `x` isn't initialized in all conditions
   --> $DIR/match-cfg-fake-edges.rs:21:13
    |
+LL |     let x;
+   |         - binding declared here but left uninitialized
+...
 LL |             x;
-   |             ^ use of possibly-uninitialized `x`
+   |             ^ `x` used here but it isn't initialized in all conditions
```
Comment (riking) [inline src/test/ui/nll/match-cfg-fake-edges.stderr:8]: I think this error is missing the assigned-here hints, because it didn't give any branch-where-missing hints.

Or it needs new text? "Assigned in unrelated match condition"

**Review by riking — COMMENTED** [review]
You should add that big test file you demonstrated as a new UI test.

File: src/test/ui/borrowck/borrowck-if-with-else.stderr [hunk 2]
```diff
@@ -1,8 +1,13 @@
-error[E0381]: use of possibly-uninitialized variable: `x`
+error[E0381]: used binding `x` isn't initialized in all conditions
   --> $DIR/borrowck-if-with-else.rs:10:9
    |
+LL |     let x: isize;
+   |         - binding declared here but left uninitialized
+LL |     if 1 > 2 {
+   |        ----- `x` is uninitialized if this condition is met
```
Comment (oli-obk) [inline src/test/ui/borrowck/borrowck-if-with-else.stderr:7]: yea, I would expect this message (or its inverse) no matter whether there's an else arm or not.

File: src/test/ui/borrowck/borrowck-init-op-equal.stderr [hunk 3]
```diff
@@ -1,8 +1,10 @@
-error[E0381]: use of possibly-uninitialized variable: `v`
+error[E0381]: used binding `v` isn't initialized
   --> $DIR/borrowck-init-op-equal.rs:3:5
    |
+LL |     let v: isize;
+   |         - binding declared here but left uninitialized
 LL |     v += 1;
-   |     ^^^^^^ use of possibly-uninitialized `v`
+   |     ^^^^^^ `v` used here but it isn't initialized
```
Comment (oli-obk) [inline src/test/ui/borrowck/borrowck-init-op-equal.stderr:7]: we should probably improve the span here at some point

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/test/ui/nll/match-cfg-fake-edges.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/nll/match-cfg-fake-edges.stderr:8]: This is interesting, it should indeed be pointing at `x = 2` in the other match arm guard.

File: src/test/ui/async-await/no-non-guaranteed-initialization.stderr [hunk 4]
```diff
@@ -1,8 +1,13 @@
-error[E0381]: use of possibly-uninitialized variable: `y`
+error[E0381]: used binding `y` isn't initialized in all conditions
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/async-await/no-non-guaranteed-initialization.stderr:1]: What about "binding `y` is used when possibly uninitialized" for the message?

I like your proposal for the if condition label, it's more straight forward to understand.

Regarding the "missing else" label, I can reword it to be similar to the other if/else labels, pointing at the `}` feels less great when rendered out, but I can try it. Pointing at the whole if could get verbose quickly, but I'm not too worried about that.

File: src/test/ui/borrowck/borrowck-if-with-else.stderr (hunk 2 above)
Comment (estebank) [reply src/test/ui/borrowck/borrowck-if-with-else.stderr:7]: Why? Because the condition's value can be statically determined?

I think what you're saying is that in the case where

```
    if y {
        x = 10;
    } else {
        println!("whoops");
    }
```

The same message can be shown as for

```
    if y {
        x = 10;
    }
```

But that is not the case because for the former we point at the ` else `, which we don't have, that's why that case is special cased.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 5]
```diff
@@ -326,6 +312,100 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+        desired_action: InitializationRequiringAction,
+        span: Span,
+        use_spans: UseSpans<'tcx>,
+    ) -> DiagnosticBuilder<'cx, ErrorGuaranteed> {
+        // We need all statements in the body where the binding was assigned to to later find all
+        // the branching code paths where the binding *wasn't* assigned to.
+        let inits = &self.move_data.init_path_map[mpi];
+        let move_path = &self.move_data.move_paths[mpi];
+        let decl_span = self.body.local_decls[move_path.place.local].source_info.span;
+        let mut spans = vec![];
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:328]: If `spans` is empty, it means that we have seen no initializer anywhere, it's just the declaration. We point at the decl and change the label slightly removing the "in some conditions", we call it "the binding is not initialized", period.

File: src/test/ui/nll/match-cfg-fake-edges.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/nll/match-cfg-fake-edges.stderr]: Figured it out. Changed it to be:

```

error[E0381]: used binding `x` isn't initialized
  --> src/test/ui/nll/match-cfg-fake-edges.rs:21:13
   |
15 |     let x;
   |         - binding declared here but left uninitialized
...
21 |             x; //~ ERROR E0381
   |             ^ `x` used here but it isn't initialized
```

Don't know if I should point at the `x = 2` in the other match arm guard 🤔

Comment (estebank) [reply src/test/ui/nll/match-cfg-fake-edges.stderr]: Changed it to be

```
error[E0381]: used binding `x` isn't initialized
  --> src/test/ui/nll/match-cfg-fake-edges.rs:21:13
   |
15 |     let x;
   |         - binding declared here but left uninitialized
...
19 |         _ if { x = 2; true } => 1,
   |                ----- binding initialized here in some conditions
20 |         _ if {
21 |             x; //~ ERROR E0381
   |             ^ `x` used here but it isn't initialized
```

File: src/test/ui/nll/issue-21232-partial-init-and-erroneous-use.stderr [hunk 6]
```diff
@@ -1,17 +1,21 @@
-error[E0381]: assign of possibly-uninitialized variable: `d`
+error[E0381]: assigned binding `d` isn't initialized
   --> $DIR/issue-21232-partial-init-and-erroneous-use.rs:28:5
    |
+LL |     let d: D;
+   |         - binding declared here but left uninitialized
 LL |     d.x = 10;
-   |     ^^^^^^^^ use of possibly-uninitialized `d`
+   |     ^^^^^^^^ `d` assigned here but it isn't initialized
```
Comment (estebank) [inline src/test/ui/nll/issue-21232-partial-init-and-erroneous-use.stderr:7]: This seems like there's a case I haven't handled.

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/98505#issuecomment-1166404448
Some tests also need a bless.
@bors r-
